mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -5484,28 +5833,110 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+          loc, GemmABlockCopyThreadClusterId_X,
+          GemmABlockCopyThreadSliceLengths_GemmKPackConstantOp);
+    } else {
+      GemmABlockCopyThreadClusterId_Y = b.create<SignedRemIOp>(
+          loc, tid, GemmABlockCopyClusterLengths_GemmKConstantOp);
+      GemmABlockCopyThreadClusterId_X = b.create<SignedDivIOp>(
+          loc, tid, GemmABlockCopyClusterLengths_GemmKConstantOp);
+      GemmAThreadDataIdBegin_Y =
+          b.create<MulIOp>(loc, GemmABlockCopyThreadClusterId_Y,
+                           GemmABlockCopyThreadSliceLengths_GemmKConstantOp);
+      GemmAThreadDataIdBegin_X =
+          b.create<MulIOp>(loc, GemmABlockCopyThreadClusterId_X,
+                           GemmABlockCopyThreadSliceLengths_GemmMConstantOp);
+    }
+
+    Value GemmAThreadDataIdBegin_Y_i32;
+    Value GemmAThreadDataIdBegin_X_i32;
+
+    Value GemmAThreadDataIdBegin_Z_i32;
+    if (KPack > 1) {
+      GemmAThreadDataIdBegin_Z_i32 = b.create<IndexCastOp>(
+          loc, GemmAThreadDataIdBegin_Z, b.getIntegerType(32));
+    }
+    GemmAThreadDataIdBegin_Y_i32 = b.create<IndexCastOp>(
         loc, GemmAThreadDataIdBegin_Y, b.getIntegerType(32));
-    auto GemmAThreadDataIdBegin_X_i32 = b.create<IndexCastOp>(
+    GemmAThreadDataIdBegin_X_i32 = b.create<IndexCastOp>(
         loc, GemmAThreadDataIdBegin_X, b.getIntegerType(32));
 
-    auto GemmABlockCopySourceCoord_Y_i32 =
-        b.create<AddIOp>(loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Y_i32);
-    auto GemmABlockCopySourceCoord_X_i32 = b.create<AddIOp>(
-        loc, m_block_data_on_global_i32, GemmAThreadDataIdBegin_X_i32);
-    auto GemmABlockCopyDestCoord_Y_i32 =
+    Value GemmABlockCopySourceCoord_Y_i32;
+    Value GemmABlockCopySourceCoord_X_i32;
+
+    Value GemmABlockCopySourceCoord_Z_i32;
+    if (KPack > 1) {
+      GemmABlockCopySourceCoord_Z_i32 = b.create<AddIOp>(
+          loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Z_i32);
+      GemmABlockCopySourceCoord_Y_i32 = b.create<AddIOp>(
+          loc, m_block_data_on_global_i32, GemmAThreadDataIdBegin_Y_i32);
+      GemmABlockCopySourceCoord_X_i32 = b.create<AddIOp>(
+          loc, zeroConstantI32Op, GemmAThreadDataIdBegin_X_i32);
+    } else {
+      GemmABlockCopySourceCoord_Y_i32 = b.create<AddIOp>(
+          loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Y_i32);
+      GemmABlockCopySourceCoord_X_i32 = b.create<AddIOp>(
+          loc, m_block_data_on_global_i32, GemmAThreadDataIdBegin_X_i32);
+    }
+
+    Value GemmABlockCopyDestCoord_Y_i32;
+    Value GemmABlockCopyDestCoord_X_i32;
+
+    Value GemmABlockCopyDestCoord_Z_i32;
+    if (KPack > 1) {
+      GemmABlockCopyDestCoord_Z_i32 = b.create<AddIOp>(
+          loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Z_i32);
+    }
+    GemmABlockCopyDestCoord_Y_i32 =
         b.create<AddIOp>(loc, zeroConstantI32Op, GemmAThreadDataIdBegin_Y_i32);
-    auto GemmABlockCopyDestCoord_X_i32 =
+    GemmABlockCopyDestCoord_X_i32 =
```

Comment:
Can we get rid of these +0 ops? They'll make the code clearer and might enable more optimizations (see #440).